File: src/chgrp.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <dirent.h>
#include <errno.h>
#include <grp.h>
#include <getopt.h>

/*
 * chgrp — change file group ownership
 *
 * Usage: chgrp [-R] [-v] [-c] GROUP FILE...
 *   -R  apply recursively
 *   -v  verbose
 *   -c  report only when a change is made
 */

static int flag_recursive = 0;
static int flag_verbose   = 0;
static int flag_changes   = 0;

/* ---- group resolution ---------------------------------------------------- */

/* Resolve a group name or numeric string to a gid.
 * Returns 0 on success, 1 on error. */
static int resolve_group(const char *spec, gid_t *gid)
{
    char *end;
    long v = strtol(spec, &end, 10);
    if (*end == '\0') {
        *gid = (gid_t)v;
        return 0;
    }
    struct group *gr = getgrnam(spec);
    if (!gr) {
        fprintf(stderr, "chgrp: invalid group: '%s'\n", spec);
        return 1;
    }
    *gid = gr->gr_gid;
    return 0;
}

/* ---- core operation ------------------------------------------------------ */

static int do_chgrp(const char *path, gid_t gid)
{
    struct stat st;
    if (lstat(path, &st) != 0) {
        fprintf(stderr, "chgrp: cannot stat '%s': %s\n", path, strerror(errno));
        return 1;
    }

    gid_t old_gid = st.st_gid;

    if (lchown(path, (uid_t)-1, gid) != 0) {
        fprintf(stderr, "chgrp: changing group of '%s': %s\n",
                path, strerror(errno));
        return 1;
    }

    if (flag_verbose || (flag_changes && old_gid != gid)) {
        if (old_gid != gid)
            printf("changed group of '%s' from %u to %u\n", path, old_gid, gid);
        else if (flag_verbose)
            printf("group of '%s' retained as %u\n", path, old_gid);
    }
    return 0;
}

static int chgrp_tree(const char *path, gid_t gid);

static int descend_dir(const char *path, gid_t gid)
{
    DIR *d = opendir(path);
    if (!d) {
        fprintf(stderr, "chgrp: cannot open directory '%s': %s\n",
                path, strerror(errno));
        return 1;
    }
    int rc = 0;
    struct dirent *de;
    while ((de = readdir(d)) != NULL) {
        if (strcmp(de->d_name, ".") == 0 || strcmp(de->d_name, "..") == 0)
            continue;
        char child[4096];
        snprintf(child, sizeof(child), "%s/%s", path, de->d_name);
        rc |= chgrp_tree(child, gid);
    }
    closedir(d);
    return rc;
}

static int chgrp_tree(const char *path, gid_t gid)
{
    int rc = do_chgrp(path, gid);
    struct stat st;
    if (lstat(path, &st) == 0 && S_ISDIR(st.st_mode))
        rc |= descend_dir(path, gid);
    return rc;
}

/* ---- main ---------------------------------------------------------------- */

static void usage(const char *prog)
{
    fprintf(stderr, "Usage: %s [-R] [-v] [-c] GROUP FILE...\n", prog);
    exit(1);
}

int main(int argc, char *argv[])
{
    if (getuid() != 0) {
        fprintf(stderr, "chgrp: must be run as root.\n");
        exit(1);
    }

    int opt;
    while ((opt = getopt(argc, argv, "Rvc")) != -1) {
        switch (opt) {
        case 'R': flag_recursive = 1; break;
        case 'v': flag_verbose   = 1; break;
        case 'c': flag_changes   = 1; break;
        default:  usage(argv[0]);
        }
    }

    if (optind + 1 >= argc)
        usage(argv[0]);

    gid_t gid;
    if (resolve_group(argv[optind++], &gid) != 0)
        exit(1);

    int rc = 0;
    for (; optind < argc; optind++) {
        if (flag_recursive)
            rc |= chgrp_tree(argv[optind], gid);
        else
            rc |= do_chgrp(argv[optind], gid);
    }
    return rc;
}

File: login_tools/cli/chgrp.py
"""
chgrp: Change file group ownership.

Usage: chgrp [-R] [-v] [-c] GROUP FILE...
"""
from __future__ import annotations

import argparse
import os
import sys
from pathlib import Path

from login_tools.db.group import GroupDb
from login_tools.privilege import require_root


def _resolve_group(group_str: str) -> int:
    """
    Resolve a group name or numeric GID string to an integer GID.
    Raises ValueError for unknown group names.
    """
    try:
        return int(group_str)
    except ValueError:
        pass
    gdb = GroupDb()
    entry = gdb.get_by_name(group_str)
    if entry is None:
        raise ValueError(f'invalid group: {group_str!r}')
    return entry.gid


def _chgrp_path(path: Path, gid: int, verbose: bool, changes: bool) -> int:
    """Apply chgrp to a single path. Returns 0 on success, 1 on error."""
    try:
        st = path.stat()
        old_gid = st.st_gid
        os.chown(path, -1, gid)
        if verbose or (changes and old_gid != gid):
            if old_gid != gid:
                print(f"changed group of '{path}' from {old_gid} to {gid}")
            elif verbose:
                print(f"group of '{path}' retained as {old_gid}")
        return 0
    except OSError as exc:
        print(f'chgrp: {path}: {exc}', file=sys.stderr)
        return 1


def _chgrp_recursive(path: Path, gid: int, verbose: bool, changes: bool) -> int:
    rc = _chgrp_path(path, gid, verbose, changes)
    if path.is_dir():
        for child in path.iterdir():
            rc |= _chgrp_recursive(child, gid, verbose, changes)
    return rc


def main() -> None:
    require_root()

    parser = argparse.ArgumentParser(
        prog='chgrp',
        description='Change file group ownership.',
    )
    parser.add_argument('group', metavar='GROUP', help='New group name or GID')
    parser.add_argument('files', nargs='+', metavar='FILE', help='Files to modify')
    parser.add_argument(
        '-R', '--recursive',
        action='store_true',
        help='Operate on files and directories recursively',
    )
    parser.add_argument(
        '-v', '--verbose',
        action='store_true',
        help='Output a diagnostic for every file processed',
    )
    parser.add_argument(
        '-c', '--changes',
        action='store_true',
        help='Like --verbose but report only when a change is made',
    )
    args = parser.parse_args()

    try:
        gid = _resolve_group(args.group)
    except ValueError as exc:
        print(f'chgrp: {exc}', file=sys.stderr)
        sys.exit(1)

    rc = 0
    for f in args.files:
        p = Path(f)
        if not p.exists() and not p.is_symlink():
            print(
                f"chgrp: cannot access '{f}': No such file or directory",
                file=sys.stderr,
            )
            rc = 1
            continue
        if args.recursive:
            rc |= _chgrp_recursive(p, gid, args.verbose, args.changes)
        else:
            rc |= _chgrp_path(p, gid, args.verbose, args.changes)
    sys.exit(rc)


if __name__ == '__main__':
    main()
